File: autocrm/notion.py
"""Drain the local outbox into Notion (match party_id, monotonic Last contacted)."""

from __future__ import annotations

import json
import logging
import os
import time
import urllib.error
import urllib.request
from collections import defaultdict
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from autocrm import outbox
from autocrm.common import DIRECTION_INBOUND, DIRECTION_OUTBOUND, OUTBOX_DB_PATH
from autocrm.outbox import OutboxRow

LOG = logging.getLogger(__name__)


def notion_configured() -> bool:
    token = os.environ.get("NOTION_TOKEN", "").strip()
    db_id = os.environ.get("NOTION_DATABASE_ID", "").strip()
    return bool(token and db_id)


def load_notion_config_from_env() -> dict[str, str]:
    token = os.environ.get("NOTION_TOKEN", "").strip()
    db_id = os.environ.get("NOTION_DATABASE_ID", "").strip()
    if not token or not db_id:
        raise RuntimeError(
            "Set NOTION_TOKEN and NOTION_DATABASE_ID"
        )
    return {
        "NOTION_TOKEN": token,
        "NOTION_DATABASE_ID": db_id,
        "PHONES_PROP": os.environ.get("NOTION_PHONES_PROP", "Phones"),
        "EMAILS_PROP": os.environ.get("NOTION_EMAILS_PROP", "Emails"),
        "LAST_CONTACTED_PROP": os.environ.get(
            "NOTION_LAST_CONTACTED_PROP", "Last Contacted"
        ),
        "LAST_CHANNEL_PROP": os.environ.get(
            "NOTION_LAST_CHANNEL_PROP", "Last Channel"
        ),
    }


def phone_digits(p: str) -> str:
    return "".join(c for c in p if c.isdigit())


def phone_match_variants(p: str) -> set[str]:
    """
    Forms used to compare handles across formatting styles.
    e.g. +1 (703) 395-5764, 7033955764, +17033955764
    """
    digits = phone_digits(p)
    if not digits:
        return set()
    variants = {digits}
    if len(digits) == 11 and digits.startswith("1"):
        variants.add(digits[1:])
    elif len(digits) == 10:
        variants.add("1" + digits)
    return variants


def normalize_email(e: str) -> str:
    return e.strip().lower()


def platform_to_channel_option(platform: str) -> str:
    """Map outbox platform to a Notion select option name."""
    if platform.strip().lower() == "text":
        return "Text"
    return platform.strip()


def channel_property_payload(platform: str, existing_prop: dict | None) -> dict:
    """Build Notion property value for Last Channel (select or legacy rich_text)."""
    label = platform_to_channel_option(platform)
    prop_type = (existing_prop or {}).get("type", "select")
    if prop_type == "rich_text":
        return {
            "rich_text": [
                {"type": "text", "text": {"content": label[:2000]}},
            ],
        }
    return {"select": {"name": label}}


def party_contact_keys(party_id: str) -> tuple[set[str], set[str]]:
    p = party_id.strip()
    if not p:
        return set(), set()
    if "@" in p:
        return set(), {normalize_email(p)}
    return phone_match_variants(p), set()


@dataclass(frozen=True)
class PageUpdate:
    page_id: str
    platform: str
    occurred_at: datetime
    row_ids: tuple[int, ...]


def plan_page_updates(
    rows: list[OutboxRow],
    pages: list[dict],
    cfg_map: Mapping[str, str],
) -> tuple[list[PageUpdate], list[int]]:
    delete_ids: list[int] = []
    by_page_rows: dict[str, list[OutboxRow]] = defaultdict(list)

    for row in rows:
        if row.direction not in (DIRECTION_INBOUND, DIRECTION_OUTBOUND):
            delete_ids.append(row.id)
            continue
        page_id = match_page_for_party(row.party_id, pages, cfg_map)
        if not page_id:
            delete_ids.append(row.id)
            continue
        by_page_rows[page_id].append(row)

    updates: list[PageUpdate] = []
    for page_id, page_rows in by_page_rows.items():
        best = max(page_rows, key=lambda r: r.created_at)
        updates.append(
            PageUpdate(
                page_id=page_id,
                platform=best.platform,
                occurred_at=_unix_to_utc(best.created_at),
                row_ids=tuple(r.id for r in page_rows),
            )
        )
    return updates, delete_ids


def _notion_headers(token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {token}",
        "Notion-Version": "2022-06-28",
        "Content-Type": "application/json",
    }


def _notion_request(
    method: str,
    url: str,
    token: str,
    body: dict | None = None,
    timeout: int = 30,
) -> dict:
    data = None
    if body is not None:
        data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        method=method,
        headers=_notion_headers(token),
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8", errors="replace")
        LOG.error("Notion HTTP %s: %s", e.code, err_body)
        raise


def _parse_notion_datetime(val: dict | None) -> datetime | None:
    if not val or val.get("type") != "date":
        return None
    d = val.get("date")
    if not d or not d.get("start"):
        return None
    start = d["start"]
    if len(start) == 10:
        start = start + "T00:00:00+00:00"
    dt = datetime.fromisoformat(start.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def _unix_to_utc(ts: float) -> datetime:
    return datetime.fromtimestamp(ts, tz=timezone.utc)


def _page_phones_emails(
    page: dict,
    phones_prop: str,
    emails_prop: str,
) -> tuple[set[str], set[str]]:
    props = page.get("properties") or {}
    phones: set[str] = set()
    emails: set[str] = set()
    p = props.get(phones_prop) or {}
    if p.get("type") == "multi_select":
        for opt in p.get("multi_select") or []:
            phones |= phone_match_variants(opt.get("name", ""))
    elif p.get("type") == "rich_text":
        for t in p.get("rich_text") or []:
            phones |= phone_match_variants(t.get("plain_text", ""))
    elif p.get("type") == "phone_number":
        if p.get("phone_number"):
            phones |= phone_match_variants(str(p["phone_number"]))

    e = props.get(emails_prop) or {}
    if e.get("type") == "email":
        if e.get("email"):
            emails.add(normalize_email(str(e["email"])))
    elif e.get("type") == "multi_select":
        for opt in e.get("multi_select") or []:
            emails.add(normalize_email(opt.get("name", "")))
    elif e.get("type") == "rich_text":
        for t in e.get("rich_text") or []:
            emails.add(normalize_email(t.get("plain_text", "")))
    return phones, emails


def match_page_for_party(
    party_id: str,
    pages: list[dict],
    cfg_map: Mapping[str, str],
) -> str | None:
    hint_phones, hint_emails = party_contact_keys(party_id)
    if not hint_phones and not hint_emails:
        return None
    phones_prop = cfg_map["PHONES_PROP"]
    emails_prop = cfg_map["EMAILS_PROP"]
    for page in pages:
        pp, ee = _page_phones_emails(page, phones_prop, emails_prop)
        if hint_phones & pp or hint_emails & ee:
            return page["id"]
    return None


def _fetch_all_pages(database_id: str, token: str) -> list[dict]:
    pages: list[dict] = []
    cursor = None
    base = "https://api.notion.com/v1/databases/" + database_id + "/query"
    while True:
        body: dict = {"page_size": 100}
        if cursor:
            body["start_cursor"] = cursor
        r = _notion_request("POST", base, token, body)
        pages.extend(r.get("results") or [])
        if not r.get("has_more"):
            break
        cursor = r.get("next_cursor")
    return pages


def _patch_page(
    page_id: str,
    occurred_at: datetime,
    platform: str,
    token: str,
    cfg_map: Mapping[str, str],
) -> None:
    last_prop = cfg_map["LAST_CONTACTED_PROP"]
    chan_prop = cfg_map["LAST_CHANNEL_PROP"]
    url = f"https://api.notion.com/v1/pages/{page_id}"
    page = _notion_request("GET", url, token)
    props = page.get("properties") or {}
    current = _parse_notion_datetime(props.get(last_prop))
    occ = occurred_at.astimezone(timezone.utc)
    if current is not None and occ < current:
        return
    body = {
        "properties": {
            last_prop: {
                "date": {"start": occ.strftime("%Y-%m-%dT%H:%M:%S.000Z")},
            },
            chan_prop: channel_property_payload(platform, props.get(chan_prop)),
        }
    }
    _notion_request("PATCH", url, token, body)


def sync_outbox(
    db_path: Path | None = None,
    *,
    cfg: dict[str, str] | None = None,
    min_interval: float | None = None,
    batch_size: int = 50,
) -> dict[str, int]:
    if min_interval is None:
        min_interval = float(os.environ.get("NOTION_MIN_INTERVAL", "0.35"))
    cfg = cfg or load_notion_config_from_env()
    token = cfg["NOTION_TOKEN"]
    db_id = cfg["NOTION_DATABASE_ID"]
    cfg_map = {
        "PHONES_PROP": cfg["PHONES_PROP"],
        "EMAILS_PROP": cfg["EMAILS_PROP"],
        "LAST_CONTACTED_PROP": cfg["LAST_CONTACTED_PROP"],
        "LAST_CHANNEL_PROP": cfg["LAST_CHANNEL_PROP"],
    }

    db = db_path or OUTBOX_DB_PATH
    outbox.init_db(db)

    all_pages = _fetch_all_pages(db_id, token)
    LOG.info("Notion sync: loaded %d pages", len(all_pages))

    applied = 0
    errors = 0

    while True:
        batch = outbox.fetch_outbox_batch(batch_size, db_path=db)
        if not batch:
            break

        updates, delete_ids = plan_page_updates(batch, all_pages, cfg_map)

        for update in updates:
            try:
                _patch_page(
                    update.page_id,
                    update.occurred_at,
                    update.platform,
                    token,
                    cfg_map,
                )
                delete_ids.extend(update.row_ids)
                applied += 1
                time.sleep(min_interval)
            except Exception:
                LOG.exception("Notion update failed for page %s", update.page_id)
                errors += 1

        outbox.delete_outbox_rows(delete_ids, db_path=db)

    return {"applied": applied, "errors": errors}

File: internal/notion/notion.go
// Package notion drains the local outbox into Notion.
package notion

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/akpiya/autocrm/internal/common"
	"github.com/akpiya/autocrm/internal/outbox"
)

// PageUpdate is a planned Notion PATCH for one page.
type PageUpdate struct {
	PageID     string
	Platform   string
	OccurredAt time.Time
	RowIDs     []int64
}

// Config holds Notion credentials and property names.
type Config struct {
	Token      string
	DatabaseID string
	Props      map[string]string
}

// ValidationResult describes whether a Notion database is usable by AutoCRM.
type ValidationResult struct {
	MissingProperties []string
}

// Configured reports whether Notion env vars are set.
func Configured() bool {
	return strings.TrimSpace(os.Getenv("NOTION_TOKEN")) != "" &&
		strings.TrimSpace(os.Getenv("NOTION_DATABASE_ID")) != ""
}

// LoadConfigFromEnv loads Notion config or returns an error.
func LoadConfigFromEnv() (Config, error) {
	token := strings.TrimSpace(os.Getenv("NOTION_TOKEN"))
	dbID := strings.TrimSpace(os.Getenv("NOTION_DATABASE_ID"))
	if token == "" || dbID == "" {
		return Config{}, fmt.Errorf("set NOTION_TOKEN and NOTION_DATABASE_ID")
	}
	return Config{
		Token:      token,
		DatabaseID: dbID,
		Props:      common.NotionPropertyConfig(),
	}, nil
}

// ValidateDatabase checks that the configured Notion database is reachable and
// exposes the properties AutoCRM needs.
func ValidateDatabase(client *http.Client, cfg Config) (ValidationResult, error) {
	if client == nil {
		client = &http.Client{Timeout: 30 * time.Second}
	}
	props := cfg.Props
	if props == nil {
		props = common.NotionPropertyConfig()
	}
	url := "https://api.notion.com/v1/databases/" + cfg.DatabaseID
	db, err := notionRequest(client, http.MethodGet, url, cfg.Token, nil)
	if err != nil {
		return ValidationResult{}, err
	}
	rawProps, _ := db["properties"].(map[string]any)
	required := []string{
		props["PHONES_PROP"],
		props["EMAILS_PROP"],
		props["LAST_CONTACTED_PROP"],
		props["LAST_CHANNEL_PROP"],
	}
	var missing []string
	for _, name := range required {
		if _, ok := rawProps[name]; !ok {
			missing = append(missing, name)
		}
	}
	return ValidationResult{MissingProperties: missing}, nil
}

// PhoneDigits strips non-digits from a phone string.
func PhoneDigits(p string) string {
	var b strings.Builder
	for _, c := range p {
		if c >= '0' && c <= '9' {
			b.WriteRune(c)
		}
	}
	return b.String()
}

// PhoneMatchVariants returns digit forms used for phone comparison.
func PhoneMatchVariants(p string) map[string]struct{} {
	digits := PhoneDigits(p)
	out := map[string]struct{}{}
	if digits == "" {
		return out
	}
	out[digits] = struct{}{}
	if len(digits) == 11 && digits[0] == '1' {
		out[digits[1:]] = struct{}{}
	} else if len(digits) == 10 {
		out["1"+digits] = struct{}{}
	}
	return out
}

// NormalizeEmail lowercases and trims an email.
func NormalizeEmail(e string) string {
	return strings.ToLower(strings.TrimSpace(e))
}

// PlatformToChannelOption maps outbox platform to a Notion select name.
func PlatformToChannelOption(platform string) string {
	switch strings.ToLower(strings.TrimSpace(platform)) {
	case "text":
		return "Text"
	case "phone_call":
		return "Phone"
	case "facetime_audio", "facetime_video":
		return "Facetime"
	default:
		return strings.TrimSpace(platform)
	}
}

// ChannelPropertyPayload builds Notion property value for Last Channel.
func ChannelPropertyPayload(platform string, existingProp map[string]any) map[string]any {
	label := PlatformToChannelOption(platform)
	propType, _ := existingProp["type"].(string)
	if propType == "rich_text" {
		return map[string]any{
			"rich_text": []map[string]any{
				{"type": "text", "text": map[string]any{"content": truncate(label, 2000)}},
			},
		}
	}
	return map[string]any{"select": map[string]any{"name": label}}
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// PartyContactKeys returns phone and email match keys for a party_id.
func PartyContactKeys(partyID string) (phones map[string]struct{}, emails map[string]struct{}) {
	p := strings.TrimSpace(partyID)
	phones = map[string]struct{}{}
	emails = map[string]struct{}{}
	if p == "" {
		return phones, emails
	}
	if strings.Contains(p, "@") {
		emails[NormalizeEmail(p)] = struct{}{}
		return phones, emails
	}
	return PhoneMatchVariants(p), emails
}

// MatchPageForParty finds a Notion page id for party_id.
func MatchPageForParty(partyID string, pages []map[string]any, cfg map[string]string) string {
	hp, he := PartyContactKeys(partyID)
	if len(hp) == 0 && len(he) == 0 {
		return ""
	}
	phonesProp := cfg["PHONES_PROP"]
	emailsProp := cfg["EMAILS_PROP"]
	for _, page := range pages {
		pp, ee := pagePhonesEmails(page, phonesProp, emailsProp)
		if intersects(hp, pp) || intersects(he, ee) {
			if id, ok := page["id"].(string); ok {
				return id
			}
		}
	}
	return ""
}

func intersects(a, b map[string]struct{}) bool {
	for k := range a {
		if _, ok := b[k]; ok {
			return true
		}
	}
	return false
}

func pagePhonesEmails(page map[string]any, phonesProp, emailsProp string) (phones, emails map[string]struct{}) {
	phones = map[string]struct{}{}
	emails = map[string]struct{}{}
	props, _ := page["properties"].(map[string]any)
	if props == nil {
		return phones, emails
	}
	if p, ok := props[phonesProp].(map[string]any); ok {
		mergePhones(p, phones)
	}
	if e, ok := props[emailsProp].(map[string]any); ok {
		mergeEmails(e, emails)
	}
	return phones, emails
}

func mergePhones(p map[string]any, phones map[string]struct{}) {
	switch p["type"] {
	case "multi_select":
		if opts, ok := p["multi_select"].([]any); ok {
			for _, o := range opts {
				if m, ok := o.(map[string]any); ok {
					for k := range PhoneMatchVariants(fmt.Sprint(m["name"])) {
						phones[k] = struct{}{}
					}
				}
			}
		}
	case "rich_text":
		if opts, ok := p["rich_text"].([]any); ok {
			for _, o := range opts {
				if m, ok := o.(map[string]any); ok {
					for k := range PhoneMatchVariants(fmt.Sprint(m["plain_text"])) {
						phones[k] = struct{}{}
					}
				}
			}
		}
	case "phone_number":
		if n := p["phone_number"]; n != nil {
			for k := range PhoneMatchVariants(fmt.Sprint(n)) {
				phones[k] = struct{}{}
			}
		}
	}
}

func mergeEmails(e map[string]any, emails map[string]struct{}) {
	switch e["type"] {
	case "email":
		if em := e["email"]; em != nil {
			emails[NormalizeEmail(fmt.Sprint(em))] = struct{}{}
		}
	case "multi_select":
		if opts, ok := e["multi_select"].([]any); ok {
			for _, o := range opts {
				if m, ok := o.(map[string]any); ok {
					emails[NormalizeEmail(fmt.Sprint(m["name"]))] = struct{}{}
				}
			}
		}
	case "rich_text":
		if opts, ok := e["rich_text"].([]any); ok {
			for _, o := range opts {
				if m, ok := o.(map[string]any); ok {
					emails[NormalizeEmail(fmt.Sprint(m["plain_text"]))] = struct{}{}
				}
			}
		}
	}
}

// PlanPageUpdates groups outbox rows into per-page updates and unmatched delete ids.
func PlanPageUpdates(rows []outbox.Row, pages []map[string]any, cfg map[string]string) ([]PageUpdate, []int64) {
	var deleteIDs []int64
	byPage := map[string][]outbox.Row{}

	for _, row := range rows {
		if row.Direction != common.DirectionInbound && row.Direction != common.DirectionOutbound {
			deleteIDs = append(deleteIDs, row.ID)
			continue
		}
		pageID := MatchPageForParty(row.PartyID, pages, cfg)
		if pageID == "" {
			deleteIDs = append(deleteIDs, row.ID)
			continue
		}
		byPage[pageID] = append(byPage[pageID], row)
	}

	var updates []PageUpdate
	for pageID, pageRows := range byPage {
		best := pageRows[0]
		for _, r := range pageRows[1:] {
			if r.CreatedAt > best.CreatedAt {
				best = r
			}
		}
		var rowIDs []int64
		for _, r := range pageRows {
			rowIDs = append(rowIDs, r.ID)
		}
		updates = append(updates, PageUpdate{
			PageID:     pageID,
			Platform:   best.Platform,
			OccurredAt: unixToUTC(best.CreatedAt),
			RowIDs:     rowIDs,
		})
	}
	return updates, deleteIDs
}

func unixToUTC(ts float64) time.Time {
	sec := int64(ts)
	nsec := int64((ts - float64(sec)) * 1e9)
	return time.Unix(sec, nsec).UTC()
}

func parseNotionDatetime(val map[string]any) *time.Time {
	if val == nil || val["type"] != "date" {
		return nil
	}
	d, _ := val["date"].(map[string]any)
	if d == nil {
		return nil
	}
	start, _ := d["start"].(string)
	if start == "" {
		return nil
	}
	if len(start) == 10 {
		start += "T00:00:00+00:00"
	}
	start = strings.Replace(start, "Z", "+00:00", 1)
	t, err := time.Parse(time.RFC3339, start)
	if err != nil {
		// try without timezone
		t, err = time.Parse("2006-01-02T15:04:05.000-07:00", start)
		if err != nil {
			return nil
		}
	}
	utc := t.UTC()
	return &utc
}

func notionRequest(client *http.Client, method, url, token string, body map[string]any) (map[string]any, error) {
	var r io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		r = bytes.NewReader(b)
	}
	req, err := http.NewRequest(method, url, r)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Notion-Version", "2022-06-28")
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		log.Printf("Notion HTTP %d: %s", resp.StatusCode, string(data))
		return nil, fmt.Errorf("notion http %d", resp.StatusCode)
	}
	var out map[string]any
	if len(data) > 0 {
		if err := json.Unmarshal(data, &out); err != nil {
			return nil, err
		}
	}
	return out, nil
}

func fetchAllPages(client *http.Client, databaseID, token string) ([]map[string]any, error) {
	var pages []map[string]any
	var cursor string
	base := "https://api.notion.com/v1/databases/" + databaseID + "/query"
	for {
		body := map[string]any{"page_size": 100}
		if cursor != "" {
			body["start_cursor"] = cursor
		}
		r, err := notionRequest(client, http.MethodPost, base, token, body)
		if err != nil {
			return nil, err
		}
		if results, ok := r["results"].([]any); ok {
			for _, p := range results {
				if m, ok := p.(map[string]any); ok {
					pages = append(pages, m)
				}
			}
		}
		if more, _ := r["has_more"].(bool); !more {
			break
		}
		if c, ok := r["next_cursor"].(string); ok {
			cursor = c
		} else {
			break
		}
	}
	return pages, nil
}

func patchPage(client *http.Client, pageID string, occurredAt time.Time, platform, token string, cfg map[string]string, cachedPage map[string]any) (bool, error) {
	lastProp := cfg["LAST_CONTACTED_PROP"]
	chanProp := cfg["LAST_CHANNEL_PROP"]
	url := "https://api.notion.com/v1/pages/" + pageID
	props := map[string]any{}
	if cachedPage != nil {
		if p, ok := cachedPage["properties"].(map[string]any); ok {
			props = p
		}
	}
	if len(props) == 0 {
		page, err := notionRequest(client, http.MethodGet, url, token, nil)
		if err != nil {
			return false, err
		}
		if p, ok := page["properties"].(map[string]any); ok {
			props = p
		}
	}
	var existingChan map[string]any
	if c, ok := props[chanProp].(map[string]any); ok {
		existingChan = c
	}
	current := parseNotionDatetime(asMap(props[lastProp]))
	occ := occurredAt.UTC()
	if current != nil && occ.Before(*current) {
		return false, nil
	}
	body := map[string]any{
		"properties": map[string]any{
			lastProp: map[string]any{
				"date": map[string]any{"start": occ.Format("2006-01-02T15:04:05.000Z")},
			},
			chanProp: ChannelPropertyPayload(platform, existingChan),
		},
	}
	if _, err := notionRequest(client, http.MethodPatch, url, token, body); err != nil {
		return false, err
	}
	return true, nil
}

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return nil
}

// SyncStats is returned from SyncOutbox.
type SyncStats struct {
	Applied int
	Errors  int
	Pending int
}

// SyncOutbox drains the outbox into Notion.
func SyncOutbox(dbPath string, cfg *Config, minInterval float64) (SyncStats, error) {
	if minInterval == 0 {
		minInterval = common.NotionMinInterval
	}
	c := Config{}
	if cfg != nil {
		c = *cfg
	} else {
		var err error
		c, err = LoadConfigFromEnv()
		if err != nil {
			return SyncStats{}, err
		}
	}
	if dbPath == "" {
		var err error
		dbPath, err = common.OutboxDBPath()
		if err != nil {
			return SyncStats{}, err
		}
	}
	cfgMap := common.NotionPropertyConfig()
	client := &http.Client{Timeout: 30 * time.Second}

	t0 := time.Now()
	pending, err := outbox.FetchAll(dbPath)
	if err != nil {
		return SyncStats{}, err
	}
	if len(pending) == 0 {
		log.Printf("Notion sync: no pending rows (%.2fs)", time.Since(t0).Seconds())
		return SyncStats{}, nil
	}

	allPages, err := fetchAllPages(client, c.DatabaseID, c.Token)
	if err != nil {
		return SyncStats{}, err
	}
	pagesByID := map[string]map[string]any{}
	for _, p := range allPages {
		if id, ok := p["id"].(string); ok {
			pagesByID[id] = p
		}
	}

	updates, deleteIDs := PlanPageUpdates(pending, allPages, cfgMap)

	applied, skipped, errors := 0, 0, 0
	var mu sync.Mutex
	lastStart := time.Time{}
	workers := common.NotionPatchWorkers
	if workers < 1 {
		workers = 1
	}

	sem := make(chan struct{}, workers)
	var wg sync.WaitGroup
	var resultMu sync.Mutex
	var toDelete []int64
	toDelete = append(toDelete, deleteIDs...)

	for _, update := range updates {
		update := update
		wg.Add(1)
		sem <- struct{}{}
		go func() {
			defer wg.Done()
			defer func() { <-sem }()

			mu.Lock()
			if minInterval > 0 && !lastStart.IsZero() {
				wait := minInterval - time.Since(lastStart).Seconds()
				if wait > 0 {
					time.Sleep(time.Duration(wait * float64(time.Second)))
				}
			}
			lastStart = time.Now()
			mu.Unlock()

			patched, err := patchPage(
				client, update.PageID, update.OccurredAt, update.Platform,
				c.Token, cfgMap, pagesByID[update.PageID],
			)
			resultMu.Lock()
			defer resultMu.Unlock()
			if err != nil {
				errors++
				return
			}
			toDelete = append(toDelete, update.RowIDs...)
			if patched {
				applied++
			} else {
				skipped++
			}
		}()
	}
	wg.Wait()

	if err := outbox.DeleteRows(toDelete, dbPath); err != nil {
		return SyncStats{}, err
	}

	log.Printf(
		"Notion sync timing: %.2fs total | pending=%d pages=%d updates=%d applied=%d skipped=%d errors=%d",
		time.Since(t0).Seconds(), len(pending), len(allPages), len(updates), applied, skipped, errors,
	)
	return SyncStats{Applied: applied, Errors: errors, Pending: len(pending)}, nil
}
